### PR TITLE
Use better types from the CLI handlers to the internals.

### PR DESCRIPTION
**Pull request by scruffystuffs**
Summary:

* Wrap the `Text` of the API key as early as possible, unwrap it as late as possible.
* Abstract building the Authorization header (where we actually need the api key text)
* Collapse project {name, revision, branch} by leveraging the inferred project and an override project, and merging them, using the resulting merge type everywhere else.

### CODE REVIEW

File: src/App/Fossa/FossaAPIV1.hs [hunk 1]
```diff
@@ -74,6 +73,9 @@ renderLocatorUrl :: Int -> Locator -> Text
 renderLocatorUrl orgId Locator{..} =
   locatorFetcher <> "+" <> T.pack (show orgId) <> "/" <> locatorProject <> "$" <> fromMaybe "" locatorRevision
 
+apiHeader :: ApiKey -> Option scheme
+apiHeader key = header "Authorization" ("token " <> encodeUtf8 (unApiKey key))
```
Comment (scruffystuffs) [inline src/App/Fossa/FossaAPIV1.hs:77]: This seems like a safe abstraction at this point.

File: src/App/Fossa/Main.hs [hunk 2]
```diff
@@ -27,38 +28,47 @@ appMain = do
 
   maybeApiKey <- checkAPIKey optAPIKey
 
+  let override = OverrideProject { overrideName = optProjectName, overrideRevision = optProjectRevision, overrideBranch = Nothing }
```
Comment (scruffystuffs) [inline src/App/Fossa/Main.hs:31]: I don't like this being on one line.  But I was getting syntax errors trying to move it around. 🤷

File: src/App/Fossa/Main.hs [hunk 3]
```diff
@@ -27,38 +28,47 @@ appMain = do
...
+requireKey :: Maybe ApiKey -> IO ApiKey
 requireKey (Just key) = pure key
 requireKey Nothing = die "A FOSSA API key is required to run this command"
 
 changeDir :: FilePath -> IO ()
 changeDir path = validateDir path >>= setCurrentDir
 
+infixr 1 <$$>
+(<$$>) :: (Functor f, Functor g) => (a -> b) -> f (g a) -> f (g b)
+(<$$>) = fmap . fmap
```
Comment (scruffystuffs) [inline src/App/Fossa/Main.hs:64]: Should this be somewhere else?  Also, should we reconsider making it an operator vs naming it?

File: src/App/Fossa/ProjectInference.hs [hunk 4]
```diff
@@ -24,9 +26,16 @@ import Text.GitConfig.Parser (Section (..), parseConfig)
 import Text.Megaparsec (errorBundlePretty)
 import Effect.ReadFS
 
+mergeOverride :: OverrideProject -> InferredProject -> ProjectRevision
+mergeOverride OverrideProject {..} InferredProject {..} = ProjectRevision name revision branch
+  where
+    name = fromMaybe inferredName overrideName
+    revision = fromMaybe inferredRevision overrideRevision
+    branch = fromMaybe inferredBranch overrideBranch
```
Comment (scruffystuffs) [inline src/App/Fossa/ProjectInference.hs:34]: @cnr There are cases where we have the branch, but never actually use it.  Does this break the laziness we had by previously not having access to it at all?

File: src/App/Fossa/Main.hs (hunk 2 above)
Comment (cnr) [reply src/App/Fossa/Main.hs:31]: you can make it multi-line if you indent the `{` (or whatever else you're breaking on) past the `o` in `override`:
```hs
let override = OverrideProject
      { overrideName = optProjectName
      , overrideRevision = optProjectRevision
      , overrideBranch = Nothing
      }
```

and then running ormolu, haha

File: src/App/Fossa/Main.hs (hunk 3 above)
Comment (cnr) [reply src/App/Fossa/Main.hs:64]: This is fine here -- also this is a nit: it should be `infixl 4` like `<$>` (see: `:info (<$>)` in ghci)

File: src/App/Fossa/Main.hs [hunk 5]
```diff
@@ -27,38 +28,47 @@ appMain = do
...
+
 -- | Try to fetch FOSSA_API_KEY from env if not supplied from cmdline
-checkAPIKey :: Maybe Text -> IO (Maybe Text)
-checkAPIKey Nothing = fmap T.pack <$> lookupEnv "FOSSA_API_KEY"
-checkAPIKey key = return key
+checkAPIKey :: Maybe Text -> IO (Maybe ApiKey)
+checkAPIKey key = case key of
+  Just key' -> pure . Just $ ApiKey key'
+  Nothing -> ApiKey <$$> T.pack <$$> lookupEnv "FOSSA_API_KEY"
```
Comment (cnr) [inline src/App/Fossa/Main.hs:70]: this will have to change slightly with `infixl 4`, but uniformity with `<$>` would help readability

Comment (scruffystuffs) [reply src/App/Fossa/Main.hs:70]: It took me way too many tries to figure out the left side becomes a pure pipeline applied to the functor.   I originally tried to do this, but I never figured it out.

**Pull request merged**